### PR TITLE
2938 - Fixing FFIConstantHandleType externalType wrong in 64 bit

### PR DESCRIPTION
**Pull request by StephanEggermont**
Getting inspired by the nice code that was written in squeak monticello repository.
Thx @eliotmiranda and @VincentBlondeau and @astares and @nicolas-cellier-aka-nice ! 

I.e.:

FFI-Kernel-eem.57
Time: 6 November 2018, 6:39:32 pm

Provide the longLong to complement the set of other integer conveniences.

FFI-Kernel-eem.55
Time: 23 August 2018, 2:12:03 pm

Add an errorCode to ExternalFunction>>invokeWithArguments:.  At least for now the design for the catching of exceptions during FFI calls has failing the callout on exception enabled by the callout having an error code.

FFI-Kernel-nice.52
Time: 29 April 2018, 6:39:52 pm

Fix pointerSpec for 64bits FFI
A pointer has a Smalltalk wordSize byte-length, rather than hardcoded 4 byte length.

FFI-Kernel-nice.49
Time: 13 April 2018, 3:56:28 pm

Give a chance to recompile the structure compiledSpec when resuming on a different platform. See ExternalStructure class>>#install.

FFI-Kernel-nice.48
Time: 13 April 2018, 8:00:08 am

Correct a bug for 64bits pointer arithmetic: unsignedLongAt: and unsignedLongAt:put: are not machine dependent, they allways fetch/store an uint32_t

No need to invoke self class wordSize, when self size already encodes the same information.

### CODE REVIEW

**Comment by StephanEggermont** [comment]
Fixes #2938

**Comment by StephanEggermont** [comment]
Needs backport

**Review by estebanlm — CHANGES_REQUESTED** [review]
mmm. while I agree this changes are useful, the right fix in FFIConstantHandleType is not to hack about the length of long, but to consider them a datatype of the same size of a pointer (which is what they are). 
Also, this harvested changes overrides some categorisations/comments we made, so it needs to be made more carefully.

**Comment by VincentBlondeau** [comment]
Tests that are failing:

```
Nouveaux échecs - 5
windows-32 / Tests-windows-32 / testImageContainsOneSubscribedOrganizer – Windows32.RPackage.Tests.RPackageWithDoTest<1s
windows-32 / Tests-windows-32 / testAliasCompositions – Windows32.Tests.Traits.TraitCompositionTest<1s
windows-32 / Tests-windows-32 / testSumWithParenthesis – Windows32.Tests.Traits.TraitCompositionTest<1s
osx-32 / Tests-osx-32 / testImageContainsOneSubscribedOrganizer – MacOSX32.RPackage.Tests.RPackageWithDoTest<1s
osx-32 / Tests-osx-32 / testAliasCompositions – MacOSX32.Tests.Traits.TraitCompositionTest<1s
Échecs préexistant - 6
windows-32 / Tests-windows-32 / testImageContainsOneSubscribedOrganizer – Windows32.RPackage.Tests.RPackageWithDoTest<1s
windows-32 / Tests-windows-32 / testAliasCompositions – Windows32.Tests.Traits.TraitCompositionTest<1s
windows-32 / Tests-windows-32 / testSumWithParenthesis – Windows32.Tests.Traits.TraitCompositionTest<1s
osx-32 / Tests-osx-32 / testImageContainsOneSubscribedOrganizer – MacOSX32.RPackage.Tests.RPackageWithDoTest<1s
osx-32 / Tests-osx-32 / testAliasCompositions – MacOSX32.Tests.Traits.TraitCompositionTest<1s
unix-32 / Tests-unix-32 / testChangeWidgetTextUpdatesPresenter – Unix32.Spec.MorphicBackendTests.MorphicTextInputFieldAdapterTest<1s
unix-32 / Tests-unix-32 / testMaxLengthIsSetInWidget – Unix32.Spec.MorphicBackendTests.MorphicTextInputFieldAdapterTest<1s
windows-32 / Tests-windows-32 / testChangeWidgetTextUpdatesPresenter – Windows32.Spec.MorphicBackendTests.MorphicTextInputFieldAdapterTest<1s
windows-32 / Tests-windows-32 / testMaxLengthIsSetInWidget – Windows32.Spec.MorphicBackendTests.MorphicTextInputFieldAdapterTest<1s
osx-32 / Tests-osx-32 / testChangeWidgetTextUpdatesPresenter – MacOSX32.Spec.MorphicBackendTests.MorphicTextInputFieldAdapterTest<1s
osx-32 / Tests-osx-32 / testMaxLengthIsSetInWidget – MacOSX32.Spec.MorphicBackendTests.MorphicTextInputFieldAdapterTest
```

**Comment by VincentBlondeau** [comment]
The failures are not related to the issue.

**Comment by estebanlm** [comment]
let me see the CI again

**Comment by guillep** [comment]
Same comment as in https://github.com/pharo-project/pharo/pull/2945

> Overall, harvesting the changes is good, but still I do not understand how they actually fix the original bug with FFIConstantHandleType. I am reading now
> 
> ```smalltalk
> FFIConstantHandleType >> externalType
> 	^ ExternalType ulong
> ```
> And I wonder: in windows HANDLEs are just pointers, why not redefine that as
> 
> ```smalltalk
> FFIConstantHandleType >> externalType
> 	^ ExternalType void asPointer
> ```
> ?
> 
> That would make much much clearer the intention.
> In any case, making that change without harvesting the changes seems futile because this seems buggy in the current version. In 64 bits I get:
> 
> ```smalltalk
> ExternalType void asPointerType byteSize => 4
> ```
> !!!

I understand that HANDLEs in windows are "integer like" but they are and they work as pointers. Making them integers does not help in anything but pretty printing. Actually it's the other way around, it complicates the calculation of types and sizes and makes people more confused...

I've been in the couple of months playing with different windows APIs and treating HANDLEs as pointers works just fine and produces no further trouble.

Maybe @tesonep has something to add...

**Comment by StephanEggermont** [comment]
I'm unsure about the next steps

**Comment by estebanlm** [comment]
let see how the build goes now :)

**Comment by guillep** [comment]
@StephanEggermont I've overseen this PR before because the state was still in "requested changes", something related to categorizations. Is that important or not? @estebanlm ?

**Comment by StephanEggermont** [comment]
I fixed the categorizations. The other parts I’m confused about

**Comment by estebanlm** [comment]
what other parts?
It seems that is ok... this symbol to decide when regenerate looks bad, but is like that in FFI-Kernel :(

**Comment by estebanlm** [comment]
I will try to integrate this, but I do not like the failing tests. Running again :)

File: src/Spec-Core/MenuItemPresenter.class.st [hunk 1]
```diff
@@ -134,12 +134,12 @@ MenuItemPresenter >> name: aString [
 { #category : #api }
 MenuItemPresenter >> performMenuActionWith: aMenuItem [
```
Comment (estebanlm) [inline src/Spec-Core/MenuItemPresenter.class.st]: why there is this change here?

**Comment by estebanlm** [comment]
Still, I don't know what `MenuItemPresenter` is doing in this PR :)

**Comment by VincentBlondeau** [comment]
@estebanlm That is why I created a new one without.

**Comment by VincentBlondeau** [comment]
See https://github.com/pharo-project/pharo/pull/3252

**Comment by estebanlm** [comment]
So we close this one :)

**Pull request closed without merging**